mobile/native/src/health.rs

```diff
@@ -22,44 +28,96 @@ pub enum ServiceStatus {
-                    result = orderbook_rx.changed() => {
-                        match result {
-                            Ok(()) => {
-                                let status = orderbook_rx.borrow();
-                                event::publish(&EventInternal::ServiceHealthUpdate((Service::Orderbook, *status)));
-                            }
-                            Err(_) => {
-                                tracing::error!("Sender dropped");
-                                event::publish(&EventInternal::ServiceHealthUpdate((Service::Orderbook, ServiceStatus::Unknown)));
-                                break;
-                            },
-                        }
-                    },
-                }
-            }
-        });
+    pub fn new(config: Config, runtime: &Runtime) -> (Self, Tx) {
+        let (orderbook_tx, orderbook_rx) = watch::channel(ServiceStatus::Unknown);
+
+        let config: ConfigInternal = config.into();
+
+        let mut tasks = Vec::new();
+
+        let orderbook_monitoring =
+            runtime.spawn(publish_status_updates(Service::Orderbook, orderbook_rx));
+        tasks.push(orderbook_monitoring);
+
+        let (coordinator_tx, coordinator_rx) = watch::channel(ServiceStatus::Unknown);
+
+        let coordinator_health_endpoint =
+            "http://".to_owned() + &config.http_endpoint.to_string() + "/health";
+        let check_coordinator = tokio::spawn(check_health_endpoint(
+            coordinator_health_endpoint,
+            coordinator_tx,
+        ));
+        tasks.push(check_coordinator);
+        let coordinator_monitoring =
+            runtime.spawn(publish_status_updates(Service::Coordinator, coordinator_rx));
+        tasks.push(coordinator_monitoring);
 
         (
-            Self { _monitoring_task },
+            Self { _tasks: tasks },
             Tx {
                 orderbook: orderbook_tx,
             },
         )
     }
 }
+
+/// Publishes the health status updates for a given service to the event hub
+async fn publish_status_updates(service: Service, mut rx: watch::Receiver<ServiceStatus>) {
+    loop {
+        select! {
+            result = rx.changed() => {
+                match result {
+                    Ok(()) => {
+                        let status = rx.borrow();
+                        event::publish(&EventInternal::ServiceHealthUpdate((service, *status)));
+                    }
+                    Err(_) => {
+                        tracing::error!("Sender dropped");
+                        event::publish(&EventInternal::ServiceHealthUpdate((service, ServiceStatus::Unknown)));
+                        break;
+                    },
+                }
+            },
+        }
+    }
+}
+
+/// Periodically checks the health of a given service and updates the watch channel
+async fn check_health_endpoint(endpoint: String, tx: watch::Sender<ServiceStatus>) {
+    loop {
+        let status = if send_request(&endpoint).await.is_ok() {
```

Comment:
I think there is a flaw in this logic: as long as the request returns you any status code (could be a `404`) you treat the service as online.

Reply:
using the method below (`error_for_status`) is a very convenient way of checking `reqwest::Response` - it returns an `Err` for any response in range 400...599: (and then you can still inspect the `Err` if needed, which I don't need in this case).

source: `https://docs.rs/reqwest/0.11.18/reqwest/struct.Response.html#method.error_for_status

Reply:
Oh, very nice. Sorry that I missed that.